src/client/components/localisation/darwin_robot/stories/model_visualiser.tsx

```diff
@@ -0,0 +1,101 @@
+import { computed } from 'mobx'
+import { Component } from 'react'
+import * as React from 'react'
+import { Camera } from 'three'
+import { Geometry } from 'three'
+import { Material } from 'three'
+import { PointLight } from 'three'
+import { CircleGeometry } from 'three'
+import { AxesHelper } from 'three'
+import { Box3 } from 'three'
+import { Mesh } from 'three'
+import { MeshPhongMaterial } from 'three'
+import { PerspectiveCamera } from 'three'
+import { Scene } from 'three'
+import { SpotLight } from 'three'
+import { Object3D } from 'three'
+
+import { Canvas } from './three'
+import { Three } from './three'
+import { cache } from './three.stories'
+import { cacheAndDispose } from './three.stories'
+
+export class ModelVisualiser extends Component<{ render(): Object3D; }> {
+  render() {
+    return <Three render={this.renderScene}/>
+  }
+
+  private renderScene = (canvas: Canvas) => {
+    const bbox = cache(() => this.bbox())
+    const geometry = cacheAndDispose(() => this.floorGeometry())
+    const material = cacheAndDispose(() => this.floorMaterial())
+    const floor = cache(() => this.floor(bbox(canvas), geometry(canvas), material(canvas)))
+    const helper = cache(() => this.helper())
+    const spotlight = cache(() => this.spotlight())
+    const pointlight = cache(() => this.pointlight(this.camera(canvas)))
+    const scene = cache(() => this.scene([
+      floor(canvas),
+      helper(canvas),
+      spotlight(canvas),
+      pointlight(canvas),
+      this.model,
+    ]))
+    const render = cache(() => ({ scene: scene(canvas), camera: this.camera(canvas) }))
+    return () => render(canvas)
+  }
+
+  private camera(canvas: Canvas) {
+    const camera = new PerspectiveCamera(75, canvas.width / canvas.height, 0.01, 100)
+    camera.position.set(0.4, 0.3, 0.3)
+    camera.up.set(0, 0, 1)
+    camera.lookAt(0, 0, 0)
+    return camera
+  }
+
+  private scene(children: Object3D[]) {
+    const scene = new Scene()
```

Comment:
Isn't this a memory leak?
Each time making a new scene means the renderer doesn't invalidate objects in the old scene

Reply:
Well, it's actually solved by the thing I commented about below.
Each renderer caches objects by scene/camera combination. So if you change the scene it ends up with another copy. Dispose works, but if you're constantly caching the same object it'll make a new copy for every scene/camera combo. In those cases you won't dispose (because it hasn't changed) but because the scene changed it got another copy. Or at least I think that's how it worked

Reply:
Although you should make sure that you're not creating new cameras each time too